Feature/subscription

habanera03 opened the pull request:
# 요약
### Issues : [#20](https://github.com/f-lab-edu/planB/issues/20)
1\. 커피 구독 등록

# 리뷰 가이드
## 1. 커피 구독 등록

- insert 중 오류 발생 시 전체 rollback
- ShopInfoCheck Class로 카페 영업 확인과 관련된 메서드 관리

hagu03 committed:
Update README.md

hagu03 committed:
리뷰 수정사항 반영
1. isNotavailable 오타 수정
2. ShopWorkingDay와 ShopDayoff 서비스 통합

hagu03 committed:
[Feature/subscription] 리뷰 수정사항 반영
1. isNotavailable 오타 수정
2. ShopWorkingDay와 ShopDayoff 서비스 통합

[Feature/subscription] 수정사항 반영
1. ShopInfo에서 businessId 제거

hagu03 committed:
[Feature/subscription-trigger] 수정사항 반영

1. TimePatternValidator 클래스 initialize 시 Pattern.compile(pattern) 실행

hagu03 committed:
feat: 커피 구독 등록

1. 사용자의 커피 구독 등록

Resolves: #20

hagu03 committed:
feat: 커피 구독 등록

1. 사용자의 커피 구독 등록

Resolves: #20

hagu03 committed:
리뷰 수정사항 반영
1. isNotavailable 오타 수정
2. ShopWorkingDay와 ShopDayoff 서비스 통합

f-lab-evan commented on src/main/java/com/flab/planb/common/ExceptionAdvice.java:
```diff
@@ -28,8 +28,23 @@
+    @ExceptionHandler(IllegalArgumentException.class)
+    public ResponseEntity<?> methodValidException(IllegalArgumentException exception,
+                                                  HttpServletRequest request) {
+        log.error("유효성 검증 실패 URI : " + request.getRequestURI());
```
동시에 여러 API에서 에러가 발생하면 어떤 api에서 어떤 에러 메세지가 발생했는지 확인 할 수 있을까요?

habanera03 replied:
순차적으로 찍힌다는 보장이 없어 확인하기 힘들 것 같습니다. 에러가 발생한 uri와 에러 메세지를 하나의 로그로 남기는 방법이 확실하게 파악할 수 있다고 생각됩니다.

f-lab-evan commented on src/main/java/com/flab/planb/controller/ShopInfoCheck.java:
```diff
@@ -0,0 +1,67 @@
+    public boolean isDayOff(long shopId, int target) {
+        log.debug("shop id : {}, target : {}", shopId, target);
+
+        return getShopDayoff(shopId).stream().anyMatch(o -> o.getDay() == target);
```
Day는 ValueObejct로 고려하는게 코드상 명확하고 깔끔해지겠네요. ValueObejct를 어떻게 request와 db query에서 맵핑 시킬지도 고려되어야 하겠네요.

f-lab-evan commented on src/main/java/com/flab/planb/controller/ShopInfoCheck.java:
```diff
@@ -0,0 +1,67 @@
+    public boolean isNotOpen(long shopId, String target) {
+        log.debug("shop id : {}, target : {}", shopId, target);
+
+        LocalTime targetTime = getLocalTime(target);
```
요일 마다 오픈/클로우즈 시간이 다를 수 있다는 가정으로 수정 해보시기 바랍니다.

habanera03 replied:
tbl_shop_working_day 테이블 추가해서 작업해보았습니다.

f-lab-evan commented on src/main/java/com/flab/planb/controller/ShopInfoCheck.java:
```diff
@@ -0,0 +1,67 @@
+        LocalTime startTime = getLocalTime(shopInfoDTO.getOperationStartTime());
+        LocalTime endTime = getLocalTime(shopInfoDTO.getOperationEndTime());
+
+        log.debug("operationStartTime : {}", shopInfoDTO.getOperationStartTime());
```
이 로그로 알수 있는 것이 무엇일까요

habanera03 replied:
isNotOpend의 boolean 결과값이 왜 그렇게 나왔는지 추측할 수 있는 용도로 사용했습니다.

f-lab-evan commented on src/main/java/com/flab/planb/controller/ShopInfoCheck.java:
```diff
@@ -0,0 +1,67 @@
+    }
+
+    private int getHour(String time) {
+        return Integer.parseInt(time.substring(0, 2));
```
시간 포멧을 이용해서 파싱하는 것이 보다 코드를 읽는 사람에게 이해하기 쉬운 코드를 전달한다고 생각합니다.

f-lab-evan commented on src/main/java/com/flab/planb/controller/SubscriptionController.java:
```diff
@@ -0,0 +1,117 @@
+    }
+
+    private boolean isDuplicateSubscription(SubscriptionRequest subscriptionRequest) {
+        return subscriptionService.existsDuplicateSucscription(subscriptionRequest) > 0;
```
오타가 있네요

f-lab-evan commented on src/main/java/com/flab/planb/dto/shop/ShopInfoDTO.java:
```diff
@@ -0,0 +1,49 @@
+    @Positive(message = "사업자정보 ID는 양수여야 합니다.")
+    private long businessId;
+    @Pattern(regexp = "^([01][0-9]|2[0-3])([0-5][0-9])$", message = "영업시작시간은 0000~2359까지 가능합니다.")
+    private String operationStartTime;
```
DB에서 DATE관련 type을 사용하는 것과 String type으로 사용하는 것에는 어떤 차이가 있을까요?

habanera03 replied:
초단위 값이 필요 없으므로 향후 조회 시 HH:mm 형식으로 포맷하는 함수를 사용하지 않기 위해 String type으로 처리하고자 했습니다. 하지만 확인해 본 결과 LocalTime으로 데이터를 받으면 초단위가 00일 경우(예를 들어 12:31:00) toString() 했을 때 12:31로 출력되는 것을 확인했습니다.
따라서 LocalTime으로 값을 받으면 구태여 쿼리에서 포맷함수를 사용하지 않아도 되고, 시간 비교를 위해 String을 LocalTime으로 파싱하지 않아도 된다고 판단하였습니다.
결론적으로 DB 컬럼 값을 chat(4) -> time으로 변경 및 Strimg -> LocalTime으로 수정하였습니다.

hagu03 committed:
refactor: 전체

1. ResponseEntity 생성 공통화
2. DTO 키워드 제거

Resolves: #23

hagu03 committed:
1. 리뷰 수정사항 반영
- getResponseEntity 메소드명 수정
- 들여쓰기 수정
- lookupMessageArgs 메소드에서 stream 없이 for문 사용으로 변경
- ResponseEntityBuilder 클래스 @Getter 제거

2. 수정사항
- newResponseEntityParam 메소드명 변경

hagu03 committed:
1. 리뷰 수정사항 반영
- getResponseEntity 메소드명 수정
- 들여쓰기 수정
- lookupMessageArgs 메소드에서 stream 없이 for문 사용으로 변경
- ResponseEntityBuilder 클래스 @Getter 제거

2. 수정사항
- newResponseEntityParam 메소드명 변경

hagu03 committed:
1. 리뷰 수정사항 반영
- getResponseEntity 메소드명 수정
- 들여쓰기 수정
- lookupMessageArgs 메소드에서 stream 없이 for문 사용으로 변경
- ResponseEntityBuilder 클래스 @Getter 제거

2. 수정사항
- newResponseEntityParam 메소드명 변경

hagu03 committed:
1. 리뷰 수정사항 반영
- getResponseEntity 메소드명 수정
- 들여쓰기 수정
- lookupMessageArgs 메소드에서 stream 없이 for문 사용으로 변경
- ResponseEntityBuilder 클래스 @Getter 제거

2. 수정사항
- newResponseEntityParam 메소드명 변경

hagu03 committed:
리뷰 수정사항 반영
- 들여쓰기 수정

f-lab-evan commented on src/main/java/com/flab/planb/annotation/TimePatternValidator.java:
```diff
@@ -0,0 +1,21 @@
+
+    @Override
+    public boolean isValid(LocalTime value, ConstraintValidatorContext context) {
+        return Pattern.compile(pattern).matcher(value.toString()).matches();
```
실패하는 경우가 어떤 경우가 있을까요?

habanera03 replied:
0분 30분 시간단위 체크를 위한 경우로 03:22와 같은 시간을 거르기 위함입니다.

f-lab-evan commented on src/main/java/com/flab/planb/controller/ShopInfoCheck.java:
```diff
@@ -0,0 +1,54 @@
+    private final ShopDayoffService shopDayoffService;
+    private final ShopWorkingDayService shopWorkingDayService;
+
+    public boolean isNotavailable(Subscription subscription) {
```
오타입니다

f-lab-evan commented on src/main/java/com/flab/planb/controller/ShopInfoCheck.java:
```diff
@@ -0,0 +1,54 @@
+    }
+
+    private boolean isNotExist(Subscription subscription) {
+        getShopInfo(subscription.getShopId());
```
다른 함수는 false를 던지는 해당 메소드는 exception을 던지고 있습니다. 다른 메소드와 같은 레벨에서 체크하는게 맞을까요?

habanera03 replied:
해당 메소드 자체가 검증을 위한 조건으로 사용하는데 애매함이 있다고 판단됩니다. getShopInfo로 얻은 데이터를 사용하는 곳이 없기 때문에 해당 메소드를 제거하고, 명확하게 존재여부를 확인할 수 있도록 exists 메소드를 새로 만들어 사용하도록 수정하였습니다.

f-lab-evan commented on src/main/java/com/flab/planb/controller/ShopInfoCheck.java:
```diff
@@ -0,0 +1,54 @@
+        log.debug("shop id : {}, target day : {}", subscription.getShopId(), subscription.getSubscriptionDay());
+
+        return getShopDayoff(subscription.getShopId())
+            .stream().anyMatch(o -> o.getDay().getCode() == subscription.getSubscriptionDay());
```
off day 확인을 db에서 select하면서도 가능할 것 같은데 수정해보시기 바랍니다.

f-lab-evan commented on src/main/java/com/flab/planb/controller/ShopInfoCheck.java:
```diff
@@ -0,0 +1,54 @@
+        return shopInfoService.findByShopId(shopId);
+    }
+
+    private List<ShopDayoff> getShopDayoff(long shopId) {
```
오타입니다.

f-lab-evan commented on src/main/java/com/flab/planb/controller/ShopInfoCheck.java:
```diff
@@ -0,0 +1,54 @@
+
+    private final ShopInfoService shopInfoService;
+    private final ShopDayoffService shopDayoffService;
+    private final ShopWorkingDayService shopWorkingDayService;
```
shopworkingday나 shopdayoff의 경우는 shop이라는 도메인의 aggregate 영역에 포함된 모델이라고 생각됩니다. 
각각의 로직이 복잡하지 않은 상태에서 굳이 각각의 서비스를 나누는 것 보다 하나의 aggregate로 처리하는게 좋아 보이는데 고민해보시기 바랍니다.

f-lab-evan commented on src/main/java/com/flab/planb/controller/SubscriptionController.java:
```diff
@@ -0,0 +1,95 @@
+
+    private void saveSubscriptionMenus(List<SubscriptionMenu> menus) {
+        subscriptionService.saveSubscriptionMenus(menus);
+        menus.forEach(m -> log.debug("subscription id : {}, menu id : {}", m.getSubscriptionId(), m.getMenuId()));
```
아직 id관련해서는 수정이 되지 않은 것으로 보입니다

habanera03 replied:
기존에 auto_increment id를 사용하는 방식에서 tbl_shop_info의 id와 tbl_sho_menu의 id를 키쌍으로 사용하도록 테이블 변경을 진행했습니다. 이로써 tbl_subscription에서 생성한 id와 요청 데이터만 가지고 tbl_subscription_menu와 tbl_subscription_menu_option 테이블 insert 처리를 하도록 변경했습니다.

hagu03 committed:
[Refactor] 수정사항 반영

1. @Transactional 처리하는 메소드를 Controller -> Service로 이동

hagu03 committed:
Merge remote-tracking branch 'origin/refactor' into refactor

# Conflicts:
#	src/main/java/com/flab/planb/controller/ShopInfoCheck.java
#	src/main/java/com/flab/planb/controller/SubscriptionController.java
#	src/main/java/com/flab/planb/dto/shop/ShopDayoff.java
#	src/main/java/com/flab/planb/dto/shop/ShopWorkingDay.java
#	src/main/java/com/flab/planb/dto/subscription/Subscription.java
#	src/test/java/com/flab/planb/controller/SubscriptionControllerIntegrationTest.java

hagu03 committed:
[Feature/subscription-trigger] 리뷰 수정사항 반영

1. test/resources/mapper 폴더 제거
2. where 절의 IF문 제거
3. 시간조회 처리를 쿼리 밖에서 처리하도록 수정

hagu03 committed:
[Feature/subscription-trigger] 수정사항 반영

1. TimePatternValidator 클래스 initialize 시 Pattern.compile(pattern) 실행

hagu03 committed:
[Feature/subscription-trigger] 수정사항 반영

1. ShedLock 사용해 멀티 서버에서 동시에 스케쥴 되는 것 방지
2. step 멀티 스레드로 처리 추가
3. 스프링 배치 테스트

hagu03 committed:
[Feature/subscription-trigger] 수정사항 반영

1. TimePatternValidator 클래스 initialize 시 Pattern.compile(pattern) 실행

hagu03 committed:
[Feature/subscription-trigger] 수정사항 반영

1. TimePatternValidator 클래스 initialize 시 Pattern.compile(pattern) 실행

hagu03 committed:
[Feature/subscription] 수정사항 반영
1. ShopInfo에서 businessId 제거

hagu03 committed:
[Feature/subscription] 수정사항 반영
1. ShopInfo에서 businessId 제거
2. ShopInfoCheck 위치 변경
3. Test 코드 정리

hagu03 committed:
[Feature/subscription] 수정사항 반영
1. ShopInfo에서 businessId 제거
2. ShopInfoCheck 위치 변경

hagu03 committed:
[Feature/subscription] 수정사항 반영
1. ShopInfo에서 businessId 제거
2. ShopInfoCheck 위치 변경
3. getShopInfo 메소드 제거 및 existsByShopId로 대체

hagu03 committed:
Feat: 커피 구독 알림

1. PushBatchConfig 멤버변수 수정
2. 알림 메시지 -> 데이터 메세지 형식으로 변경

Resolves: #21

hagu03 committed:
Merge branch 'feature/subscription-trigger' into refactor

# Conflicts:
#	src/test/resources/mapper/shopInfoMapper.xml

hagu03 committed:
Merge branch 'refactor' into feature/subscription

# Conflicts:
#	src/main/java/com/flab/planb/controller/SubscriptionController.java
#	src/main/resources/mapper/shopInfoMapper.xml
#	src/test/resources/mapper/shopInfoMapper.xml
